### PR TITLE
ROX-12233: e2e tests on image component resolver

### PR DESCRIPTION
**Pull request by keyallis**
## Description

A detailed explanation of the changes in your PR.

Feel free to remove this section if it is overkill for your PR, and the title of your PR is sufficiently descriptive.

## Checklist
- [ ] Investigated and inspected CI test results
- [ ] Unit test and regression tests added

### CODE REVIEW

**Comment by md2119** [comment]
rebase and re-request review once CI is fixed

File: central/deployment/datastore/datastore.go [hunk 1]
```diff
@@ -107,6 +108,39 @@ func New(dacky *dackbox.DackBox, keyFence concurrency.KeyFence, pool *pgxpool.Po
 	return newDataStore(storage, dacky, pool, bleveIndex, processIndex, images, baselines, networkFlows, risks, deletedDeploymentCache, processFilter, clusterRanker, nsRanker, deploymentRanker)
 }
 
+// NewTestDataStore allows for direct creation of the datastore for testing purposes
+func NewTestDataStore(storage store.Store, graphProvider graph.Provider, pool *pgxpool.Pool,
```
[the remark this answers is not in the document]
Comment (md2119) [reply central/deployment/datastore/datastore.go:112]: Once that is added, the first check in the function must be `if t == nil return err msg`.

File: central/graphql/resolvers/image_components_test.go [hunk 2]
```diff
@@ -0,0 +1,434 @@
...
+
+func (s *GraphQLImageComponentTestSuite) TestImageComponentImages() {
+	ctx := SetAuthorizerOverride(s.ctx, allow.Anonymous())
+
+	imageCompTests := []struct {
+		name        string
+		id          string
+		expectedIDs []string
+	}{
+		{
```
Comment (md2119) [inline central/graphql/resolvers/image_components_test.go:240]: would be good to test with two images with same components but different operating systems:
image.id = <<some_sha>>
image.scan.operatingSystem = <<some_os>>

image.id = <<some_other_sha>>
image.scan.operatingSystem = <<some_other_os>>

File: central/graphql/resolvers/image_components_test.go [hunk 3]
```diff
@@ -0,0 +1,434 @@
...
+			s.Equal(expectedCount, count)
+
+			counter, err := comp.ImageVulnerabilityCounter(ctx, RawQuery{})
+			s.NoError(err)
+			s.Equal(expectedCount, counter.All(ctx).Total(ctx))
+			s.Equal(test.expectedCounterValues.fixable, counter.All(ctx).Fixable(ctx))
+			s.Equal(test.expectedCounterValues.critical, counter.Critical(ctx).Total(ctx))
+			s.Equal(test.expectedCounterValues.important, counter.Important(ctx).Total(ctx))
+			s.Equal(test.expectedCounterValues.moderate, counter.Moderate(ctx).Total(ctx))
+			s.Equal(test.expectedCounterValues.low, counter.Low(ctx).Total(ctx))
```
Comment (md2119) [inline central/graphql/resolvers/image_components_test.go:353]: By now, you probably are using in couple of test cases. Consider factoring it out into a func.
`func(ctx, t, expected, actual)`

File: central/graphql/resolvers/image_components_test.go [hunk 4]
```diff
@@ -0,0 +1,434 @@
...
+		},
+	}
+
+	for _, test := range imageCompTests {
+		s.T().Run(test.name, func(t *testing.T) {
+			image := s.getImageResolver(ctx, test.id)
+			expectedCount := int32(len(test.expectedIDs))
+
+			comps, err := image.ImageComponents(ctx, PaginatedQuery{})
+			s.NoError(err)
```
Comment (md2119) [inline central/graphql/resolvers/image_components_test.go:201]: this must be `assert.NoError(t,...)`. You are running independent test cases within a suite. 

Same comment for all such check instances.

File: central/graphql/resolvers/image_vulnerabilities_test.go [hunk 5]
```diff
@@ -72,6 +72,8 @@ type GraphQLImageVulnerabilityTestSuite struct {
 
 func (s *GraphQLImageVulnerabilityTestSuite) SetupSuite() {
 
+	s.T().Skip("Skipping this test suite until it is revisited as the data has changed due to operating systems being added (ROX-12231)")
```
Comment (md2119) [inline central/graphql/resolvers/image_vulnerabilities_test.go:75]: Don't disable the test. Create a `testImagesWithOperatingSystem()` for image component tests instead.

File: central/graphql/resolvers/image_components_test.go [hunk 6]
```diff
@@ -391,44 +454,43 @@ func (s *GraphQLImageComponentTestSuite) TestImageComponentDeployments() {
...
-			idList := getIDList(ctx, deps)
-			s.ElementsMatch(test.expectedIDs, idList)
+			assert.NoError(s.T(), err)
+			assert.Equal(t, expectedCount, int32(len(deps)))
+			assert.ElementsMatch(s.T(), test.expectedIDs, getIDList(ctx, deps))
 
 			count, err := comp.DeploymentCount(ctx, RawQuery{})
-			s.NoError(err)
-			s.Equal(expectedCount, count)
+			assert.NoError(s.T(), err)
```
Comment (md2119) [inline central/graphql/resolvers/image_components_test.go:478]: `s.T()` -> `t`

**Review by md2119 — APPROVED** [review]

File: central/graphql/resolvers/test_utils.go [hunk 7]
```diff
@@ -573,3 +657,41 @@ func getNodeVulnerabilityResolver(ctx context.Context, t *testing.T, resolver *R
...
+	return imageComponentDataStore.New(nil, imageCompStore, imageCompIndexer, imageCompSearcher, risks, ranking.NewRanker())
+}
+func getImageCVEEdgeDatastore(ctx context.Context, db *pgxpool.Pool, gormDB *gorm.DB) imageCVEEdgeDataStore.DataStore {
+	imageCveEdgeStore := imageCVEEdgePostgres.CreateTableAndNewStore(ctx, db, gormDB)
+	imageCveEdgeIndexer := imageCVEEdgePostgres.NewIndexer(db)
+	imageCveEdgeSearcher := imageCVEEdgeSearch.NewV2(imageCveEdgeStore, imageCveEdgeIndexer)
+	return imageCVEEdgeDataStore.New(nil, imageCveEdgeStore, imageCveEdgeSearcher)
+}
+
+func getImageComponentCVEEdgeDatastore(ctx context.Context, db *pgxpool.Pool, gormDB *gorm.DB) componentCVEEdgeDataStore.DataStore {
```
Comment (md2119) [inline central/graphql/resolvers/test_utils.go:687]: Refer to `central/graphql/resolvers/testutils.go`

Comment (keyallis) [reply central/graphql/resolvers/test_utils.go:687]: I can work on collapsing the `testutils.go` file with the `test_utils.go` file in some future refactor

**Pull request merged**